src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -74,77 +71,113 @@ ov::Tensor Bank::get(int64_t uid, const std::string& device) {
-
-        ov::parallel_for(vec.size(), [&](std::size_t idx) {
-            const auto& lt = vec[idx];
-            std::unique_lock dev_guard(device_bank.mutex);
-            auto iter_device_registered = device_bank.registered_tensors.find(lt);
-            NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
-                        "Tensor should be registered first!");
-            if (device_bank.storage[iter_device_registered->second].tensor) {
-                // Already allocated
-                return;
-            }
-            dev_guard.unlock();
-
-            // Allocation and/or evaluation needed
-            // Evaluate concurrently, lock the device
-            // mutex only to update the device bank (& allocate on-device memory, if needed)
-            const auto& transformed_tensor = lt.eval();
-
-            dev_guard.lock();
-            if (device_for_alloc == "CPU") {
-                // No allocation needed
-                device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = transformed_tensor;
-                const_cast<LazyTensor&>(lt).detach();
-                return;
-            }
 
-            ov::SoPtr<ov::ITensor> remote_tensor;
-            ov::Tensor allocated_tensor;
+        guard.unlock();
+
+        if (device_for_alloc == "CPU") {
+            evaluate_cpu(device_bank, to_process);
+        } else {
+            evaluate_and_allocate_on_device(device_bank, to_process, device_for_alloc);
+        }
+    }
+}
+
+void Bank::evaluate_cpu(Bank::DeviceBank& device_bank, const std::vector<LazyTensor>& to_process) {
+    std::unique_lock guard(m_mutex);
+
+    ov::parallel_for(to_process.size(), [&](std::size_t idx) {
+        const auto& lt = to_process[idx];
+        auto iter_device_registered = device_bank.registered_tensors.find(lt);
+        NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                    "Tensor should be registered first!");
+        auto uid = iter_device_registered->second;
+        device_bank.storage.at(uid).tensor = lt.eval();
+        const_cast<LazyTensor&>(lt).detach();
+    });
+}
+
+void Bank::evaluate_and_allocate_on_device(Bank::DeviceBank& device_bank,
+                                           const std::vector<LazyTensor>& to_process,
+                                           const std::string& device) {
+    std::unique_lock guard(m_mutex);
+
+    std::vector<TensorToAllocate> uids_to_allocated(uid_count);
+
+    // Allocation needed
+    for (std::size_t idx = 0; idx < to_process.size(); ++idx) {
+        const auto& lt = to_process[idx];
+        auto iter_device_registered = device_bank.registered_tensors.find(lt);
+        NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                    "Tensor should be registered first!");
+        auto uid = iter_device_registered->second;
+        uids_to_allocated[uid] = {lt.eval_meta(), ov::Tensor(), true};
```

Comment:
here you only insert `true`-s

Reply:
Because initially it's all `false`

Reply:
Reworked again. Added uid field to TensorToAllocate + sorting based on uid